### PR TITLE
Switch from `npm` to `yarn`

### LINKED ISSUE #6879: Lockfile churn
The `package-lock.json` file changes upon every `npm install`, even when no changes are made to dependencies. This lockfile churn has caused problems for many contributors, and makes adding new dependencies without including this cruft rather tedious and time consuming.

Eliminating this churn would make development a great deal easier.

### PR DESCRIPTION
**Pull request by Gudahtt**
As a solution to the constant lockfile churn issues we've had with `npm`, the project now uses `yarn` to manage dependencies.

The `package-lock.json` file has been replaced with `yarn.lock`, which was created using `yarn migrate`. It should approximate the contents of `package-lock.json` fairly well, though there may be some changes due to deduplication. The codeowners file has been updated to reference this new lockfile.

All documentation and npm scripts have been updated to reference `yarn` rather than `npm`. Note that running scripts using `npm run` still works fine, but it seemed better to switch those to `yarn` as well to avoid confusion.

The `npm-audit` Bash script has been replaced with `yarn-audit`. The output of `yarn audit` is a bit different than `npm audit` in that it returns a bitmask to describe which severity issues were found. This
made it simpler to check the results directly from the Bash script, so the associated `npm-audit-check.js` script was no longer required. The output should be exactly the same, and the information is still sourced from the same place (the npm registry).

The new `yarn-audit` script does have an external dependency: `jq`. However, `jq` is already assumed to be present by another CI script, and is present on all CI images we use. `jq` was not added to `package.json` as a dependency because there is no official package on the npm registry, just wrapper scripts. We don't need it anywhere exept on CI anyway.

The section in `CONTRIBUTING` about how to develop inside the `node_modules` folder was removed, as the advice was a bit dated, and wasn't specific to this project anyway.

Fixes #6879

### CODE REVIEW

File: development/mock-dev.js [hunk 1]
```diff
@@ -9,7 +9,7 @@
  * This is a convenient way to develop and test the plugin
  * without having to re-open the plugin or even re-build it.
  *
- * To use, run `npm run mock`.
+ * To use, run `yarn mock`.
```
Comment (Gudahtt) [inline development/mock-dev.js]: This command doesn't exist anymore 🤔 

It looks like it was removed in this commit: https://github.com/MetaMask/metamask-extension/commit/b6789a6e2fd9168e0bf8d386611d4c439ccd9e18

I'll put up a separate PR to remove this file as well, given that it can presumably no longer be used.

File: .circleci/scripts/yarn-audit [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+yarn audit --level moderate
+audit_status="$?"
+
+# Use a bitmask to ignore INFO and LOW severity audit results
+# See here: https://yarnpkg.com/lang/en/docs/cli/audit/
+audit_status="$(( audit_status & 11100 ))"
+
+if [[ "$audit_status" != 0 ]]
+then
+    count="$(yarn audit --json | tail -1 | jq '.data.vulnerabilities.moderate + .data.vulnerabilities.high + .data.vulnerabilities.critical')"
```
Comment (Gudahtt) [inline .circleci/scripts/yarn-audit]: This still isn't counting the correct thing - it needs to discard dev dependencies. I'll fix this.

File: development/ui-dev.js [hunk 3]
```diff
@@ -12,7 +12,7 @@
  * without having to re-enter your password
  * every time the plugin rebuilds.
  *
- * To use, run `npm run ui`.
+ * To use, run `yarn ui`.
```
Comment (Gudahtt) [inline development/ui-dev.js]: Same case here as above - this can be deleted.

File: docs/notices.md [hunk 4]
```diff
@@ -2,14 +2,14 @@
 
 To add a notice:
 ```
-npm run generateNotice
+yarn generateNotice
```
Comment (Gudahtt) [inline docs/notices.md]: This file should be deleted in a separate PR as well - it was removed in https://github.com/MetaMask/metamask-extension/commit/ac8b56a00defff4cb44a6a34251a19d8ab6159b6

File: test/e2e/run-drizzle.sh [hunk 5]
```diff
@@ -4,28 +4,28 @@ set -e
...
-npm run ganache:start -- -b 2 >> /dev/null 2>&1 &
-npm_run_ganache_start_pid=$!
+yarn ganache:start -- -b 2 >> /dev/null 2>&1 &
+ganache_start_pid=$!
 sleep 5
 
 pushd "$(mktemp -d)"
-npm install --no-package-lock truffle
-truffle="$(npm bin)/truffle"
+yarn install --no-lockfile truffle
```
Comment (Gudahtt) [inline test/e2e/run-drizzle.sh]: This seems like something we shouldn't do, but it looks like this test isn't used anymore? Maybe it can be deleted?

File: .circleci/scripts/yarn-audit (hunk 2 above)
Comment (Gudahtt) [reply .circleci/scripts/yarn-audit]: It looks like `yarn` is incorrectly reporting our development dependencies as production dependencies in these security audits 😞. Until this is fixed, we may have to hold off in migrating to yarn.

Comment (Gudahtt) [reply .circleci/scripts/yarn-audit]: Nevermind, found a workaround! The flag `--groups dependencies` does effectively narrow the results to just the production dependencies.

There is still a bug in the advisory output... but that's not something we need to rely upon thankfully.

File: package.json [hunk 6]
```diff
@@ -9,27 +9,27 @@
...
-    "test:integration": "npm run test:integration:build && npm run test:flat",
+    "test:integration": "yarn test:integration:build && yarn test:flat",
     "test:integration:build": "gulp build:scss",
     "test:e2e:drizzle:beta": "SELENIUM_BROWSER=chrome test/e2e/run-drizzle.sh",
     "test:e2e:chrome": "SELENIUM_BROWSER=chrome test/e2e/run-all.sh",
     "test:web3:chrome": "SELENIUM_BROWSER=chrome test/e2e/run-web3.sh",
     "test:web3:firefox": "SELENIUM_BROWSER=firefox test/e2e/run-web3.sh",
     "test:e2e:firefox": "SELENIUM_BROWSER=firefox test/e2e/run-all.sh",
-    "test:coverage": "nyc --reporter=text --reporter=html npm run test:unit && npm run test:coveralls-upload",
+    "test:coverage": "nyc --reporter=text --reporter=html npm run test:unit && yarn test:coveralls-upload",
```
Comment (Gudahtt) [inline package.json]: I had to use `npm run` here instead of `yarn` to run the `test:unit` script because of a problem with how `yarn` and `spawn-wrap` (used by `nyc`) are interacting.

`yarn` is creating a shell script proxy for `node` and `yarn` to ensure the version used _within_ these scripts is the same as the version used to _invoke_ the script, which is generally what you want to happen (before this change, it relied upon the `PATH`, which could be messed up in unexpected ways in a variety of circumstances). But `nyc` specifically relies upon wrapping `node` in order to add instrumentation.

I'll link the `yarn` issue when I've created it - this needs a bit more investigation. In the meantime, using `npm run` works just fine.

**Comment by Gudahtt** [comment]
This is ready to review. I ran into a couple of problems, but they've all been resolved now.

**Comment by whymarrh** [comment]
@Gudahtt and just like that we've got conflicts 😅

File: test/e2e/run-drizzle.sh (hunk 5 above)
Comment (whymarrh) [reply test/e2e/run-drizzle.sh]: I think we can delete the Drizzle tests, since we haven't been running them for a while now

Comment (Gudahtt) [reply test/e2e/run-drizzle.sh]: I've created a separate PR for this: #6855

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -32,20 +32,20 @@ We recommend creating a folder somewhere manually called `node_modules`. For exa
...
+- Never make any meaningful edits to code inside an "npm/yarn-managed" node_modules folder (such as `~/code/node_modules/dat/node_modules`), because any changes inside those folders could inadvertently be deleted when installing an updated copy of a module. This has happened to me many times, so I just always use my git copy and delete the npm registry copy (as described above) to make edits to a module.
+- You should never need to run any npm/yarn commands in terminal when at your "manually managed"" node_modules folder at `~/code/node_modules`. Never running npm/yarn commands at that folder also your git copies of the modules from getting erased.
 - The location of your "manually managed" node_modules folder should be somewhere isolated from your normal require path. E.g. if you put it at `~/node_modules`, then when you run `npm install dat` at `~/Desktop` npm might decide to erase your git copy of dat at `~/node_modules/dat` and replace it with a copy from npm, which could make you lose work. Putting your manually managed `node_modules` folder in a sub-folder like `~/code` gets it "out of the way" and prevents accidents like that from happening.
```
Comment (rekmarks) [inline CONTRIBUTING.md]: `npm install` language still used here

**Review by rekmarks — DISMISSED** [review]
I believe that the other reviewers' opinions should carry greater weight than mine on this matter, but aside from some minor language nitpicks (see comment on `CONTRIBUTING.md`), I'd be excited to try `yarn`.

LGTM.

File: CONTRIBUTING.md (hunk 7 above)
Comment (Gudahtt) [reply CONTRIBUTING.md]: I was hesitant to remove it from this paragraph because it's discussing working on dependencies, and really when working on a dependency you should be using whichever package manager that project uses (whether it's yarn, npm, pnpm, etc.) The project being discussed in this case is `dat`, which does use `npm`.

It certainly could be explained better - the switching back and forth is a little confusing. I was hesitant to spend too much effort on this though, as this workflow likely isn't what we want to encourage anymore. `npm link` offers basically the same advantages, plus it's a more commonly used workflow and it offers slightly better guarantees in terms of dependency deduplication. The workflow described here still works just fine, but it's the older way of doing things, and I think it's a bit harder to understand. The only distinct advantage I can think of is that it avoids symlinks, which can be problematic on certain environments.

Comment (whymarrh) [reply CONTRIBUTING.md]: I would vote for removing this section outright, it doesn't quite need to be here

Comment (Gudahtt) [reply CONTRIBUTING.md]: Fair enough - I've removed it.

**Review by whymarrh — DISMISSED** [review]
LGTM! This is a first step to hopefully spending less time tinkering with npm failures.

**Review by frankiebee — DISMISSED** [review]

**Review by whymarrh — APPROVED** [review]
LGTM

File: package.json (hunk 6 above)
Comment (Gudahtt) [reply package.json]: Here is the relevant Yarn issue: https://github.com/yarnpkg/yarn/issues/6746

Comment (whymarrh) [reply package.json]: See also: #7994, where I'm updating this
